### PR TITLE
Add ibus-engines.rime

### PR DESCRIPTION
**Pull request by pmeiyu**
###### Motivation for this change

This pull request adds ibus-engines.rime.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
@pmeiyu please solve the merge conflict.

**Comment by pmeiyu** [comment]
@SuperSandro2000 I have rebased the code onto master branch.  You can merge now.  Thanks.

**Comment by pmeiyu** [comment]
@SuperSandro2000 Are you a merge conflict detection bot?  This repository is full of useless bots.

**Comment by pmeiyu** [comment]
@SuperSandro2000 Bad bot.  Waste my time.  Please notify me only when you are actually going to review the code.

**Comment by eyJhb** [comment]
Result of `nixpkgs-review pr 87735` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ibus-engines.rime</li>
  </ul>
</details>

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  src = fetchFromGitHub {
+    owner = "rime";
+    repo = "ibus-rime";
+    rev = version;
+    sha256 = "0zbajz7i18vrqwdyclzywvsjg6qzaih64jhi3pkxp7mbw8jc5vhy";
+  };
+
+  buildInputs = [ gdk-pixbuf glib ibus libnotify librime brise ];
+  nativeBuildInputs = [ cmake pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:16]:
```suggestion
  nativeBuildInputs = [ cmake pkg-config ];
```

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
+{ stdenv, fetchFromGitHub, cmake, pkgconfig, gdk-pixbuf, glib, ibus, libnotify
```
Comment (SuperSandro2000) [inline pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, cmake, pkg-config, gdk-pixbuf, glib, ibus, libnotify
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -3031,6 +3031,8 @@ in
       protobuf = pkgs.protobuf3_8.overrideDerivation (oldAttrs: { stdenv = clangStdenv; });
     };
 
+    rime = callPackage ../tools/inputmethods/ibus-engines/ibus-rime { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:3034]: I think this should be in ibus-engines a few lines above but I am not 100% sure.

Comment (pmeiyu) [reply pkgs/top-level/all-packages.nix:3038]: But this is indeed in ibus-engines.

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix (hunk 1 above)
Comment (pmeiyu) [reply pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix]: Fixed.

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix (hunk 2 above)
Comment (pmeiyu) [reply pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix]: Fixed.

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    substituteInPlace rime.xml \
+       --replace '/usr' $out
+  '';
+
+  meta = with stdenv.lib; {
+    isIbusEngine = true;
+    description = "Rime input method engine for IBus";
+    homepage = "https://rime.im/";
+    license = licenses.gpl3;
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:40]: `gpl3` is unclear, use `gpl3Plus` or `gpl3Only`. See https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  buildInputs = [ gdk-pixbuf glib ibus libnotify librime brise ];
+  nativeBuildInputs = [ cmake pkg-config ];
+
+  makeFlags = [ "PREFIX=$(out)" ];
+  dontUseCmakeConfigure = true;
+
+  prePatch = ''
+    substituteInPlace Makefile \
+       --replace 'cmake' 'cmake -DRIME_DATA_DIR=${brise}/share/rime-data'
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:23]: You should be able to specify this in `cmakeFlags`, see https://github.com/NixOS/nixpkgs/pull/77877/files#diff-22b83f2a09917d6f17fa4fb028cd36fd499c7d046298d46202d60a2fea4dde9fR39-R43

in combination with https://github.com/NixOS/nixpkgs/pull/77877#discussion_r491836081

Comment (pmeiyu) [reply pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:23]: The ibus-rime project used a weird mixture of Makefile and cmake.  I choose to use the make build system here.  Therefore I have to patch Makefile.  The other PR used cmake build system and that caused other dirt hacks.  I think using Makefile is cleaner  here.

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix (hunk 4 above)
Comment (pmeiyu) [reply pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix]: Fixed.  License is changed to gpl3Plus.

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:3038]: nvm.

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix (hunk 5 above)
Comment (jtojnar) [reply pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:23]: The issue with this is that it steps around the flags we pass to CMake like `cmakeBuildType`. I have opened https://github.com/rime/ibus-rime/pull/112 to allow us to use CMake for everything, including correct paths in the files.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 87735` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ibus-engines.rime</li>
  </ul>
</details>

File: pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix (hunk 5 above)
Comment (pmeiyu) [reply pkgs/tools/inputmethods/ibus-engines/ibus-rime/default.nix:23]: Great. So this will be resolved in next ibus-rime release.

**Pull request merged**